rclcpp/include/rclcpp/qos_overriding_options.hpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__QOS_OVERRIDING_OPTIONS_HPP_
+#define RCLCPP__QOS_OVERRIDING_OPTIONS_HPP_
+
+#include <functional>
+#include <initializer_list>
+#include <ostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/qos.hpp"
+#include "rclcpp/visibility_control.hpp"
+
+#include "rcl_interfaces/msg/set_parameters_result.hpp"
+#include "rmw/qos_policy_kind.h"
+
+namespace rclcpp
+{
+
+enum class RCLCPP_PUBLIC_TYPE QosPolicyKind
+{
+  AvoidRosNamespaceConventions = RMW_QOS_POLICY_AVOID_ROS_NAMESPACE_CONVENTIONS,
+  Deadline = RMW_QOS_POLICY_DEADLINE,
+  Depth = RMW_QOS_POLICY_DEPTH,
+  Durability = RMW_QOS_POLICY_DURABILITY,
+  History = RMW_QOS_POLICY_HISTORY,
+  Lifespan = RMW_QOS_POLICY_LIFESPAN,
+  Liveliness = RMW_QOS_POLICY_LIVELINESS,
+  LivelinessLeaseDuration = RMW_QOS_POLICY_LIVELINESS_LEASE_DURATION,
+  Reliability = RMW_QOS_POLICY_RELIABILITY,
+  Invalid = RMW_QOS_POLICY_INVALID,
+};
+
+RCLCPP_PUBLIC
+const char *
+qos_policy_kind_to_cstr(const QosPolicyKind & qpk);
+
+RCLCPP_PUBLIC
+std::ostream &
+operator<<(std::ostream & os, const QosPolicyKind & qpk);
+
+using QosCallbackResult = rcl_interfaces::msg::SetParametersResult;
+using QosCallback = std::function<QosCallbackResult(const rclcpp::QoS &)>;
+
+namespace detail
+{
+// forward declare
+template<typename T>
+class QosParameters;
+}
+
+/// Options that are passed in subscription/publisher constructor to specify QoSConfigurability.
+/**
+ * This options struct allows configuring:
+ * - Which policy kinds will have declared parameters.
+ * - An optional callback, that will be called to validate the final qos profile.
+ * - An optional id. In the case that different qos are desired for two publishers/subscriptions in
+ *   the same topic, this id will allow disambiguating them.
+ *
+ * Example parameter file:
+ *
+ * ```yaml
+ * my_node_name:
+ *  ros__parameters:
+ *    qos_overrides:
+ *      /my/topic/name:
+ *        publisher:  # publisher without provided id
+ *          reliability: reliable
+ *          depth: 100
+ *        publisher_my_id:  # publisher with `id="my_id"
+ *          reliability: reliable
+ *          depth: 10
+ * ```
+ */
+class QosOverridingOptions
```

Comment:
I just noticed that we're not consistent with the case used for the name `Qos`. In rclcpp, when we write CamelCase there's a mix of capitalizing the "s" and leaving it lowercase. I.e. `QoS` vs `Qos`. We should probably be consistent, though I'm not sure which is the preferred casing.

I'm pointing it out here since this PR introduces several instances of using a lowercase "s", `Qos`. Though there were mixed conventions even before this PR.

Reply:
In rclpy, we seem to be more consistent in using `QoS`, with the exception of changes introduced by this feature.

Reply:
My preference is `QoS`, but according to the google style guide, only the first letter of an acronym is supposed to be capitalized, so it would be `RosThing` rather than `ROSThing`, which I don't care for, but maybe that's where this is coming from.

https://google.github.io/styleguide/cppguide.html#General_Naming_Rules
> For the purposes of the naming rules below, a "word" is anything that you would write in English without internal spaces. This includes abbreviations, such as acronyms and initialisms. For names written in mixed case (also sometimes referred to as "camel case" or "Pascal case"), in which the first letter of each word is capitalized, prefer to capitalize abbreviations as single words, e.g., StartRpc() rather than StartRPC().

I definitely prefer QoS though, because otherwise it's ambiguous as to where or not the `o` would be capitalized if the acronym were to be taken out of the class name.

Reply:
I also prefer `QoS` (and `ROSThing` over `RosThing`), I think simply because that's how I would write it in other contexts. But I'd like see consistency at the very least.

Also, I think in this case it will be less disruptive to users if we change the APIs introduced in PR to match most of the other code using `QoS`, since this is something new not yet released into a ROS distro (except Rolling).

Reply:
Agreed.

Reply:
I think it's ok to modify all instances of `QosOverridingOptions` are modified to `QOSOverridingOptions` or `QoSOverridingOptions`.

We already have some instances of `QOS` in rclcpp (see [here](https://github.com/ros2/rclcpp/blob/dc7e95aaa524f758ff8ccf903698065420f9cc48/rclcpp/include/rclcpp/qos_event.hpp#L84)), so I prefer that one.

I find `QoS` weird, e.g. a CamelCase to snake case converter is going to get it wrong (qo_s).
I also find QOSOverridingOptions hard to read, but better than QoSOverridingOptions.
e.g. most CamelCase to snake_case converters will get the first one right `qos_overriding_options`, whereas the second one would be `qo_s_overriding_options`.

Reply:
> I find QoS weird, e.g. a CamelCase to snake case converter is going to get it wrong (qo_s).

That's a reasonable argument.

Though, if we're choosing between `Qos` and `QOS`, I'd rather us just follow the google style guide ([as documented](https://index.ros.org/doc/ros2/Contributing/Code-Style-Language-Versions/#id3)) and pick `Qos`, since refactoring will be disruptive either way.

Reply:
> I find QoS weird, e.g. a CamelCase to snake case converter is going to get it wrong (qo_s).

I don't feel that QoS is weird as that is the initialization outside of the class name. Whether or not a camel case to snake case converter is smart enough to handle this isn't really important imo. We only use it with the Messages to message header conversion as far as I know and even that was a bad idea in my mind. The convert is never going to be quite right. There are already other cases it fails for, there was an issue about it... but it's something like `Point3D` becomes `point_3_d`. These names are inherently contextual.

> Though, if we're choosing between Qos and QOS, I'd rather us just follow the google style guide (as documented) and pick Qos, since refactoring will be disruptive either way.

But refactoring new code is less disruptive than existing code. If we're gonna do this I think we should list all the uses of QOS, QoS, and Qos and consider which would be most disruptive for users.

Reply:
> But refactoring new code is less disruptive than existing code. If we're gonna do this I think we should list all the uses of QOS, QoS, and Qos and consider which would be most disruptive for users.

I agree whole-heartedly with this.  I think consistency with what is already there is more important than any particular spelling.